Sources/PackageGraph/DependencyResolver.swift

```diff
@@ -1094,3 +1153,139 @@ public class DependencyResolver<
         return container
     }
 }
+
+/// The resolver debugger.
+///
+/// Finds the constraints which results in graph being unresolvable.
+private struct ResolverDebugger<
+    Provider: PackageContainerProvider,
+    Delegate: DependencyResolverDelegate
+> where Provider.Container.Identifier == Delegate.Identifier {
+
+    typealias Identifier = Provider.Container.Identifier
+    typealias Constraint = PackageContainerConstraint<Identifier>
+
+    /// Reference to the resolver.
+    unowned let resolver: DependencyResolver<Provider, Delegate>
+
+    /// Create a new debugger.
+    init(_ resolver: DependencyResolver<Provider, Delegate>) {
+        self.resolver = resolver
+    }
+
+    // Returns the constraints which should be removed in order to make the
+    // graph resolvable.
+    func debug(
+        dependencies: [Constraint],
+        pins: [Constraint]
+    ) -> (dependencies: [Constraint], pins: [Constraint]) {
+        // Put the resolver in incomplete mode to avoid cloning new repositories.
+        resolver.isInIncompleteMode = true
+
+        let deltaAlgo = DeltaAlgorithm<ResolverChange>()
+        let allPackages = Set(dependencies.map({ $0.identifier }))
+
+        // Compute the set of changes.
+        let changes = dependencies.map({ ResolverChange.allowPackage($0.identifier) })
+                    + pins.map({ ResolverChange.allowPin($0.identifier) })
+
+        // Run the delta debugging algorithm.
+        let badChanges = deltaAlgo.run(changes: Set(changes)){ changes in
```

Comment:
I don't think this is quite right...

This will find the smallest set of packages which, taken by themselves, are unsatisfiable.

However, resolving that small set without the interactions of all of the other packages may lead (I think, I find this hard to reason about) to different failure modes than the one the full graph hits.

I think it would make sense to instead find the smallest set of constraints which can be removed from the complete set in order to make it satisfiable. That is, we would take the provided changes, subtract that from the list of all changes, and then check if that result is satisfiable.

It would be very interesting to try to construct a test case where this matters...

Does this make sense?